### ISSUE TITLE
[BUG] Session-based example breaks in TF 2.10

### ISSUE THREAD

**Issue by marcromeyn** [NVIDIA-Merlin/models#722]
### Bug description

More info can be found here: https://github.com/NVIDIA-Merlin/models/runs/8223278321?check_suite_focus=true

**Comment by bschifferer** [comment]
I checked and reproduced the error. I think this is related to the data loader. The error is raised in `merlin/models/loader/backend.py`. (see below)

It is interesting, that it works, if I run one batch manually through the network by using `model_bi_lstm(next(iter(train_dl))[0])`.

Using this line works:
```
model_bi_lstm(next(iter(train_dl))[0])
history = model_bi_lstm.fit(
    train_dl,
    validation_data=valid,
    batch_size=BATCH_SIZE,
    epochs=EPOCHS,
    schema=schema_model,
)
```

```
---------------------------------------------------------------------------
StopIteration                             Traceback (most recent call last)
Input In [106], in <cell line: 1>()
----> 1 history = model_bi_lstm.fit(
      2     train_dl,
      3     validation_data=valid,
      4     batch_size=BATCH_SIZE,
      5     epochs=EPOCHS,
      6     schema=schema_model,
      7 )

File /usr/local/lib/python3.8/dist-packages/merlin/models/tf/models/base.py:725, in BaseModel.fit(self, x, y, batch_size, epochs, verbose, callbacks, validation_split, validation_data, shuffle, class_weight, sample_weight, initial_epoch, steps_per_epoch, validation_steps, validation_batch_size, validation_freq, max_queue_size, workers, use_multiprocessing, train_metrics_steps, **kwargs)
    717 callbacks = self._add_metrics_callback(callbacks, train_metrics_steps)
    719 fit_kwargs = {
    720     k: v
    721     for k, v in locals().items()
    722     if k not in ["self", "kwargs", "train_metrics_steps", "__class__"]
    723 }
--> 725 return super().fit(**fit_kwargs)

File /usr/local/lib/python3.8/dist-packages/keras/utils/traceback_utils.py:70, in filter_traceback.<locals>.error_handler(*args, **kwargs)
     67     filtered_tb = _process_traceback_frames(e.__traceback__)
     68     # To get the full stack trace, call:
     69     # `tf.debugging.disable_traceback_filtering()`
---> 70     raise e.with_traceback(filtered_tb) from None
     71 finally:
     72     del filtered_tb

File /usr/local/lib/python3.8/dist-packages/merlin/models/tf/dataset.py:322, in BatchedDataset.__getitem__(self, idx)
    316 def __getitem__(self, idx):
    317     """
    318     implemented exclusively for consistency
    319     with Keras model.fit. Does not leverage
    320     passed idx in any way
    321     """
--> 322     return DataLoader.__next__(self)

File /usr/local/lib/python3.8/dist-packages/merlin/models/loader/backend.py:356, in DataLoader.__next__(self)
    355 def __next__(self):
--> 356     return self._get_next_batch()

File /usr/local/lib/python3.8/dist-packages/merlin/models/loader/backend.py:393, in DataLoader._get_next_batch(self)
    391 # try to iterate through existing batches
    392 try:
--> 393     batch = next(self._batch_itr)
    394 except StopIteration:
    395     # anticipate any more chunks getting created
    396     # if not, raise the StopIteration
    397     if not self._working and self._buff.empty:

StopIteration: 
```

**Comment by oliverholworthy** [comment]
With tensorflow 2.10, after running a model `.fit` with our `BatchedDataset` - the attributes `_workers` and `_batch_itr` remain set to non-None values. 

Previously, after iterating through all batches/epochs, these `BatchedDataset` would be reset to None values, so that a second call to `.fit` works as the notebook expected. 

This looks like it's something to do with the way that `len` method (our implementation calls the `stop` method -> resetting `_workers` and `_batch_itr` to None) is called on the dataset during the training of the keras model (`.fit`) 

One option I think is to implement the `on_epoch_end` method and call `stop` in there so that after each epoch the dataset is reset and can be iterated through again. Implemented in this PR: https://github.com/NVIDIA-Merlin/models/pull/724

**Comment by oliverholworthy** [comment]
This has been fixed by #724 

With an optional follow-up of #727 which may improve performance
